Add support for parent and child page restrictions in page_layout definition

mickenorlen opened the pull request:
# Update 8/7

See comment: https://github.com/AlchemyCMS/alchemy_cms/pull/1776#issuecomment-651109428

---

## What is this pull request for?
Add support for restricting parent/child_pages:


### Update 17/5

Renamed `sub_pages` to `child_pages` and added `include`/`exclude` as suggested by @mamhoff.
```YML
# config/alchemy/page_layouts.yml

- name: blog
  child_pages: 
     include: [blog_post] # array of allowed child_pages.

- name: blog_post
  parent_pages: 
     include: [blog] # array of parent pages under which page can be created
  child_pages: false # hides add child_page button from page tree

- name: no_blogs_pages # Excludes example
  child_pages:
    exclude: [blog] 
  parent_pages:
    exclude: [blog]
```

### Note
- `If child_pages = false` it will entirely hide the create child_page button (+) in page_tree. See screenshot 1.
- `if selectable_layouts.length == 1`  i added autoselect and removed blank option in create selection. See screenshot 2 

### Changes in the code
- Passed in the `page_layout` of the parent_page to `PageLayout.selectable_layouts` to be able to check the definitions.
- Added `allowing_child_pages: page.definition['child_pages'] != false` to the `page_tree_serializer` so i could hide the create (+) button.

### Screenshots
1. No (+) button if `child_pages = false`
![image](https://user-images.githubusercontent.com/9784434/78499699-8c1dac80-7752-11ea-8ba0-a20449b95ad8.png)

2. No blank/"Please choose" option when only 1 to choose.
![image](https://user-images.githubusercontent.com/9784434/78499751-be2f0e80-7752-11ea-9803-b426ed7a6c4a.png)

## Tests 
I wanted to check with you if the overall implementation looks good before fixing the tests. Currently 2 fails saying `undefined method page_layout for nil:NilClass`
```
rspec ./spec/features/admin/page_creation_feature_spec.rb:21 # Page creation overlay GUI when having a Page in the clipboard contains tabs for creating a new page and pasting from clipboard
rspec ./spec/requests/alchemy/admin/pages_controller_spec.rb:261 # Alchemy::Admin::PagesController with logged in editor user #new with current language present pages in clipboard should load all pages from clipboard
```

## Checklist
- [X] I have followed [Pull Request guidelines](https://github.com/AlchemyCMS/alchemy_cms/blob/master/CONTRIBUTING.md)
- [X] I have added a detailed description into each commit message
- [x] I have added tests to cover this change

mamhoff commented on app/serializers/alchemy/page_tree_serializer.rb:
```diff
@@ -56,6 +56,7 @@ def page_hash(page, has_children, level, folded)
         visible: page.visible?,
         restricted: page.restricted?,
         page_layout: page.page_layout,
+        allowing_sub_pages: page.definition['sub_pages'] != false,
```
You probably want to ask `.present?` instead here, to account for page definitions that do not have `sub_pages` set.

mickenorlen replied:
> You probably want to ask .present? instead here, to account for page definitions that do not have sub_pages set.

This feature was intended as an optional "restriction" rather than required specification. If they don't have it set I think all sub pages should be available as normal by default. Only if explicitly set it would restrict availability, and only if set to false should the button be removed.

> I have a few implementation comments, and I would like to suggest to only implement ONE of `sub_pages` or `parent_pages`. If a page layout can only have certain subpages, then that's all you should ever need, am I right?

I think it's worth it to be able to say that `blog_post` is only available under `blog` so I don't have to list all other layouts except `blog_post` on all other layouts.

Another thought I had now is that this previously used page_layout property `layout['hide']` (mentioned in `allowed_in_definition`), could be replaced by `parent_pages = false`. Meaning that no pages allow creating it => it's hidden. What do you think?

mamhoff replied:
I get it, ok. I'm still not happy with the idea of having the page_layouts array for one page having to check the definitions of all other page layouts. 

What about having an `only` and an `except` key in the page_layouts.yml: 
```yml
# config/alchemy/page_layouts.yml

- name: home
  sub_pages: 
    except: 
    - blog_post # array of allowed sub_pages.

- name: blog
  sub_pages: 
    only: 
    - blog_post # array of allowed sub_pages.

- name: blog_post
  sub_pages:
    only: []
``

mickenorlen replied:
If you look at `page_layout.rb` before my changes (below). It still needs to loop through each layout to check for uniqueness and "hide" etc. So we're already doing these cross layout conditions. And i think it is actually really convenient/useful to have the `parent_page` restriction as well so you don't have to check and repeat what to exclude on every page. I doesn't seem to cost anything to add another condition in there.
```
def selectable_layouts(language_id, only_layoutpages = false)
  @language_id = language_id
  all.select do |layout|
    if only_layoutpages
      layout['layoutpage'] && layout_available?(layout)
    else
      !layout['layoutpage'] && layout_available?(layout)
    end
  end
end

def layout_available?(layout)
  !layout['hide'] && !already_taken?(layout) && available_on_site?(layout)
end
```

But I still like your idea of exclude/include. What about this pattern (below) that would give the user all the options? (`*` = all, `!` = not)

Then we could say that the default value of `sub_pages`/`parent_pages` is `[*]` unless specified. And we could replace this `layout['hide']`/`hide: true`  option with `parent_pages: false`

```
- name: blog
  sub_pages: [blog_post] # array of allowed sub_pages.

- name: blog_post
  parent_pages: [blog] # array of parent pages under which page can be created
  sub_pages: false


- name: default
  parent_pages: [*]
  sub_pages: [*] 

- name: all_except_blog_as_sub_page
  sub_pages: [*, !blog]

- name: creatable_under_all_except_blog
  parent_pages: [*, !blog]

- name: hidden
  parent_pages: false
  # replaces hide: true
```

What do you think? Thanks

mickenorlen replied:
@mamhoff, i don't know why i started thinking about this `*` and `!` syntax. I think your version with `only`/`except` is better :)

But do you think it's ok to keep the `parent_page` option? thanks

mamhoff commented on app/views/alchemy/admin/pages/_new_page_form.html.erb:
```diff
@@ -4,8 +4,9 @@
-    include_blank: Alchemy.t('Please choose'),
+    include_blank: @page_layouts.length == 1 ? nil : Alchemy.t('Please choose'),
     required: true,
+    selected: @page_layouts.length == 1 ? @page_layouts.first : nil,
```
`page_layouts.first.presence` instead would do the trick

mickenorlen replied:
nice catch will fix next round

mickenorlen replied:
Actually I don't think i can change this as i select first only if length == 1 - Not if first is present

mamhoff commented on lib/alchemy/page_layout.rb:
```diff
@@ -83,9 +83,14 @@ def layouts_with_own_for_select(page_layout_name, language_id, only_layoutpages
+      def selectable_layouts(language_id, only_layoutpages = false, page_layout_name = nil)
         @language_id = language_id
+        @page_layout = page_layout_name.present? ? all.detect{ |layout| layout['name'] == page_layout_name } : nil
+        return [] if @page_layout.present? && @page_layout['sub_pages'] == false
```
Here as well, use `present?` to account for `nil` values

mickenorlen replied:
Again, this would mean that all page layouts are required to specify their sub_pages, while i only intended this as an optional restriction.

mamhoff replied:
Oh yeah, we wouldn't want that. This needs to be absolutely optional.

mickenorlen commented on lib/alchemy/page_layout.rb:
```diff
@@ -127,7 +133,28 @@ def human_layout_name(layout)
+          (excludes && !excludes.include?(layout['name']))
+      end
+
+      def layout_available_from_child?(layout)
```
It's not about creating a parent for a child. But under which parents the layout will be available. Previously there was a setting `hide = true` which made certain layouts hidden from creation. I just deleted this `hide` setting from the conditions in my last push as this can be achived with `layout['parent_pages'] = false.

But the reason why I think this feature makes sense to have is the following. Imagine that you have a mature website with some 15 page layouts available. Then you want a new type of blog post layout `special_blog_post` that should only be creatable under the blog layout.

With the new feature you simply set `parent_pages['only'] = ['blog']` in the `special_blog_post` settings.

Without it you would have to revisit all other page_layout and append `child_pages['except'] = [ ... 'special_blog_post']  

... Which feels a lot more tedious/error prone. In the end this logic is only executed on admin page creation, and we're already looping through all layouts to check the rest of the conditions - so i dont think it could have any significant impact on performance?

What do you think?

mickenorlen commented on lib/alchemy/page_layout.rb:
```diff
@@ -83,9 +83,15 @@ def layouts_with_own_for_select(page_layout_name, language_id, only_layoutpages
-      def selectable_layouts(language_id, only_layoutpages = false)
+      def selectable_layouts(language_id, only_layoutpages = false, page_layout_name = nil)
         @language_id = language_id
+        @page_layout = page_layout_name.present? ? all.detect{ |layout| layout['name'] == page_layout_name } : nil
```
This gets the layout of the parent page (Settings obj where we can check for child_page restrictions) from the parent layout name (String).

Then just below we filter with select and all the conditions.

mickenorlen commented on lib/alchemy/page_layout.rb:
```diff
@@ -83,9 +83,15 @@ def layouts_with_own_for_select(page_layout_name, language_id, only_layoutpages
+      #   name of current/parent page layout
       #
-      def selectable_layouts(language_id, only_layoutpages = false)
+      def selectable_layouts(language_id, only_layoutpages = false, page_layout_name = nil)
```
I'm not sure I can simplify this. I added more comments to the code to clear up any misunderstanding.

mickenorlen commented:
Thanks for your comments. I updated most of the things.

Some questions/thoughts:

- See code comments where i explain/argue for the `parent_pages` setting
- I see you prefer the term `sub_pages`. But IF you agree the `parent_pages` setting is something to keep in the end - I think `parent_pages` and `child_pages` are more naturally related terms. What do you think?  


> The selectable layouts methods are worth refactoring so that its easier to understand what is going on. In the end it should be easy as:

> All defined page layouts for that site - already taken unique ones - not allowed layouts = selectable layouts for the new page

This seems kind of like what we have?

In method `selectable_layouts` - This code is from before and I'm guessing it is still relevant? Basically we select filter through all the available layouts depending on if its a layoutpage or not...

```
  all.select do |layout|
    if only_layoutpages
      layout['layoutpage'] && layout_available?(layout)
    else
      !layout['layoutpage'] && layout_available?(layout)
    end
  end
```

Then below we have the conditions tested for each layout in the select filter. I only added the first two (and removed `hide` see code comments). 
```
  def layout_available?(layout)
    layout_available_from_parent?(layout) && layout_available_from_child?(layout) &&
      !already_taken?(layout) && available_on_site?(layout)
  end
```

Which equates to:
> All defined page layouts for that site - already taken unique ones - not allowed layouts = selectable layouts for the new page

tvdeyen commented:
Could you please extract the "No please choose if only one layout available" feature into a dedicated PR?

tvdeyen commented on app/controllers/alchemy/admin/pages_controller.rb:
```diff
@@ -57,9 +57,9 @@ def info
       def new
         @page ||= Page.new(layoutpage: params[:layoutpage] == "true", parent_id: params[:parent_id])
-        @page_layouts = PageLayout.layouts_for_select(@current_language.id, @page.layoutpage?)
+        @page_layouts = PageLayout.layouts_for_select(@current_language.id, @page.layoutpage?, @page.parent &.page_layout)
```
Shouldn't this be?

```suggestion
        @page_layouts = PageLayout.layouts_for_select(@current_language.id, @page.layoutpage?, @page.parent&.page_layout)
```

A test would have could this. Could you please add one?

tvdeyen commented on app/controllers/alchemy/admin/pages_controller.rb:
```diff
@@ -57,9 +57,9 @@ def info
+        @page_layouts = PageLayout.layouts_for_select(@current_language.id, @page.layoutpage?, @page.parent &.page_layout)
         @clipboard = get_clipboard("pages")
-        @clipboard_items = Page.all_from_clipboard_for_select(@clipboard, @current_language.id, @page.layoutpage?)
+        @clipboard_items = Page.all_from_clipboard_for_select(@clipboard, @current_language.id, @page.layoutpage?, @page.parent &.page_layout)
```
Ditto

```suggestion
        @clipboard_items = Page.all_from_clipboard_for_select(@clipboard, @current_language.id, @page.layoutpage?, @page.parent&.page_layout)
```

tvdeyen commented on app/serializers/alchemy/page_tree_serializer.rb:
```diff
@@ -55,6 +55,7 @@ def page_hash(page, has_children, level, folded)
         public: page.public?,
         restricted: page.restricted?,
         page_layout: page.page_layout,
+        child_pages_allowed: page.definition["child_pages"] != false,
```
I do not like the name of this attribute 

`allowed_page_types`, or maybe `nestable_pages` `nestable_child_pages`?

mickenorlen replied:
Now i named it for what it actually is `show_create_page_btn`. Which can be true even if there are no pages available, as the button is only hidden if `definition['child_pages'] == false`.

Also added tests for controller method calls. I've been using this feature for quite some time already, `[space] &.` works just as fine as `[no_space].&` - but i still removed the space as you like.

I think the implementation is good now and for me the feature is essential for controlling the page tree structure. It is also just as you requested it to be as explained in this comment https://github.com/AlchemyCMS/alchemy_cms/pull/1776#issuecomment-651109428

I would say it cannot be simplified more. You have to give me more concrete feedback if you still disagree.

tvdeyen reviewed (commented):
I looked at this again today and although I think this is a nice way to structure pages I still think that the page layout code is getting more and more complex. Since I need to maintain this going forward I need more time to get convinced. Maybe refactoring the code into smaller understandable classes and methods would help. But in the current form I tend to not accept this although I like that feature.

tvdeyen commented:
@mickenorlen we switched the default branch to `main`. I already updated this PR, but you want to pull the `main` branch and rebase this branch against this as well, so you get future updates. sorry for any inconvenience and thanks for the contribution.

mickenorlen commented:
@tvdeyen Force pushed a fresh single squash merge commit into main as rebase was a mess.
